engine/access/ingestion2/execution_result_container.go

```diff
@@ -0,0 +1,218 @@
+// Returns the number of receipts added (0 if receipt already exists, 1 if added).
+// CAUTION: not concurrency safe! Caller must hold a lock.
+//
+// Expected error returns during normal operations:
+//   - [ErrIncompatibleReceipt]: if the receipt's execution result is different from the container's result ID
+func (c *ExecutionResultContainer) addReceipt(receipt *flow.ExecutionReceipt) (uint, error) {
+	resultID := receipt.ExecutionResult.ID()
+	if resultID != c.resultID {
+		return 0, fmt.Errorf("receipt is for result %v, while ExecutionResultContainer pertains to result %v: %w",
+			resultID, c.resultID, ErrIncompatibleReceipt)
+	}
+
+	receiptID := receipt.ID()
+	if c.has(receiptID) {
+		return 0, nil
+	}
+	c.receipts[receiptID] = receipt.Stub()
+	return 1, nil
+}
+
+// Has returns whether a receipt with the given ID exists in the container.
+func (c *ExecutionResultContainer) Has(receiptID flow.Identifier) bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.has(receiptID)
+}
+
+// has returns whether a receipt with the given ID exists in the container.
+// CAUTION: not concurrency safe! Caller must hold a lock.
+func (c *ExecutionResultContainer) has(receiptID flow.Identifier) bool {
+	_, found := c.receipts[receiptID]
+	return found
+}
+
+// Size returns the number of receipts in the container.
+func (c *ExecutionResultContainer) Size() uint {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return uint(len(c.receipts))
+}
+
+// Result returns the execution result for this container.
+func (c *ExecutionResultContainer) Result() *flow.ExecutionResult {
+	// No locking is required here since the result is immutable after instantiation.
+	return c.result
+}
+
+// ResultID returns the ID of the execution result for this container.
+func (c *ExecutionResultContainer) ResultID() flow.Identifier {
+	// No locking is required here since the resultID is immutable after instantiation.
+	return c.resultID
+}
+
+// BlockHeader returns the header of the block executed by this result.
+func (c *ExecutionResultContainer) BlockHeader() *flow.Header {
+	// No locking is required here since the blockHeader is immutable after instantiation.
+	return c.blockHeader
+}
+
+// BlockView returns the view of the block executed by this result.
+func (c *ExecutionResultContainer) BlockView() uint64 {
+	// No locking is required here since the blockHeader is immutable after instantiation.
+	return c.blockHeader.View
+}
+
+// Pipeline returns the pipeline associated with this container.
+func (c *ExecutionResultContainer) Pipeline() optimistic_sync.Pipeline {
+	// No locking is required here since the pipeline is immutable after instantiation.
+	return c.pipeline
+}
+
+// ResultStatus returns the status of this result.
+func (c *ExecutionResultContainer) ResultStatus() ResultStatus {
+	return c.resultStatus.Value()
+}
+
+// SetResultStatus sets the status of this result.
+//
+// No error returns expected during normal operations.
```

Comment:
[optional] For expressiveness of the API, I would suggest we return a dedicated sentinel here. This allows the caller to optimistically attempt to advance the state and then handle errors if they occur, which requires a sentinel here for good style.